Add configuration needed for cassandra/micrometer integration when both features are present.

wetted opened the pull request:
see https://github.com/micronaut-projects/micronaut-cassandra/pull/377

wetted committed:
Add configuration needed for cassandra/micrometer integration when both features are present.

timyates commented on starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java:
```diff
@@ -53,6 +53,13 @@ public void apply(GeneratorContext generatorContext) {
+
+        if (generatorContext.isFeaturePresent(MicrometerFeature.class)) {
+            generatorContext.getConfiguration().put("cassandra.default.advanced.metrics.factory.class", "MicrometerMetricsFactory");
+            generatorContext.getConfiguration().put("cassandra.default.advanced.metrics.factory.class", "MicrometerMetricsFactory");
```
These 2 lines are the same?

timyates commented on starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java:
```diff
@@ -53,6 +53,13 @@ public void apply(GeneratorContext generatorContext) {
+        if (generatorContext.isFeaturePresent(MicrometerFeature.class)) {
+            generatorContext.getConfiguration().put("cassandra.default.advanced.metrics.factory.class", "MicrometerMetricsFactory");
+            generatorContext.getConfiguration().put("cassandra.default.advanced.metrics.factory.class", "MicrometerMetricsFactory");
+            generatorContext.getConfiguration().put("cassandra.default.advanced.metrics.session.enabled", "[connected-nodes, cql-requests, bytes-sent, bytes-received]");
+            generatorContext.getConfiguration().put("cassandra.default.advanced.metrics.node.enabled", "[cql-requests, bytes-sent, bytes-received]");
```
I don't think this will work when the config is in properties files.

I think the values need to be lists themselves, not string representation of lists.

Did you try creating an app using this?

timyates reviewed (changes requested):
I can't get an app created with this to start...

With the config in the properties as

```
cassandra.default.advanced.metrics.session.enabled=[connected-nodes,cql-requests,bytes-sent,bytes-received]
```

I get (after a minute or so when it tries to generate metrics)

```
Message: Error instantiating class MicrometerMetricsFactory (specified by advanced.metrics.factory.class): hardcoded value: advanced.metrics.session.enabled has type STRING rather than LIST
Path Taken: new HealthMonitorTask(CurrentHealthStatus currentHealthStatus,List healthIndicators) --> new HealthMonitorTask(CurrentHealthStatus currentHealthStatus,[List healthIndicators]) --> new CassandraHealthIndicator([CqlSession cqlSession])
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2322)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2277)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2289)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3056)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:81)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2958)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2919)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2730)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1693)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2165)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1328)
	at io.micronaut.cassandra.health.$CassandraHealthIndicator$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2307)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2277)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2289)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3056)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:81)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2958)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2919)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2893)
	at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3521)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistrations(DefaultBeanContext.java:3473)
	at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3443)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1436)
	at io.micronaut.context.AbstractBeanResolutionContext.getBeansOfType(AbstractBeanResolutionContext.java:95)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBeansOfType(AbstractInitializableBeanDefinition.java:2253)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractInitializableBeanDefinition.java:1476)
	at io.micronaut.management.health.monitor.$HealthMonitorTask$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2307)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2277)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2289)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3056)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:81)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2958)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2919)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2730)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1693)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:840)
	at io.micronaut.scheduling.processor.ScheduledMethodProcessor.lambda$process$2(ScheduledMethodProcessor.java:124)
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141)
	at io.micrometer.core.instrument.Timer.lambda$wrap$0(Timer.java:196)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalArgumentException: Error instantiating class MicrometerMetricsFactory (specified by advanced.metrics.factory.class): hardcoded value: advanced.metrics.session.enabled has type STRING rather than LIST
	at com.datastax.oss.driver.internal.core.util.Reflection.resolveClass(Reflection.java:336)
```

And if I change the config to

```
cassandra.default.advanced.metrics.session.enabled[0]=connected-nodes,
cassandra.default.advanced.metrics.session.enabled[1]=cql-requests
cassandra.default.advanced.metrics.session.enabled[2]=bytes-sent
cassandra.default.advanced.metrics.session.enabled[3]=bytes-received
```

I get an exception:

```
10:39:55.893 [scheduled-executor-thread-1] ERROR i.m.c.CassandraSessionFactory - Failed to instantiate CQL session: path parameter: Invalid path 'advanced.metrics.session.enabled[0]': Token not allowed in path expression: '[' (you can double-quote this token if you really want it here)
com.typesafe.config.ConfigException$BadPath: path parameter: Invalid path 'advanced.metrics.session.enabled[0]': Token not allowed in path expression: '[' (you can double-quote this token if you really want it here)
	at com.typesafe.config.impl.PathParser.parsePathExpression(PathParser.java:155)
	at com.typesafe.config.impl.PathParser.parsePathExpression(PathParser.java:74)
	at com.typesafe.config.impl.PathParser.parsePath(PathParser.java:61)
	at com.typesafe.config.impl.Path.newPath(Path.java:230)
	at com.typesafe.config.impl.PropertiesParser.fromPathMap(PropertiesParser.java:92)
	at com.typesafe.config.impl.ConfigImpl.fromAnyRef(ConfigImpl.java:265)
```

I wonder if this shows there's an issue in here https://github.com/micronaut-projects/micronaut-cassandra/pull/377

timyates commented:
Works (with [this comment applied](https://github.com/micronaut-projects/micronaut-starter/pull/2113#discussion_r1350056935)) with YAML config... but not properties config 🤔

wetted commented:
> Works (with [this comment applied](https://github.com/micronaut-projects/micronaut-starter/pull/2113#discussion_r1350056935)) with YAML config... but not properties config 🤔

Should work with property files now. It is a mismatch between Micronaut's and Datastax' notation for indexed java properties. I fixed that on the `micronaut-cassandra` PR.

wetted committed:
Cassandra feature cleanup, correct outdated datastax configuration and add configuration for test-resources.

wetted commented:
I have cleaned up the Cassandra feature. It was generating configuration that was outdated and incorrect. I also added necessary test-resources configuration.

For this to really work with test-resources a generated project needs 

```
micronaut {
    testResources {
        clientTimeout = 600
    }
}

```

However, there isn't an easy way to contribute that right now. I am going to work on adding this with an extension, like we do for the MicronautAot feature. I'll add that on a separate PR and sync this up after it's merged. It's not like a project generated with cassandra and test-resources features were ready to go anyway.

timyates commented on starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java:
```diff
@@ -44,15 +53,35 @@ public String getDescription() {
+            configuration.put("test-resources.containers.cassandra.startup-timeout", "600s");
+            configuration.put("test-resources.containers.cassandra.image-name", "cassandra");
+            configuration.put("test-resources.containers.cassandra.exposed-ports[0].cassandra.port", 9042);
+            configuration.put("cassandra.default.basic.contact-points[0]", "localhost:${cassandra.port}");
```
Why does this have a `[0]` in the key? Also, line 83?

timyates replied:
```suggestion
            configuration.put("cassandra.default.basic.contact-points", List.of("localhost:${cassandra.port}"));
```

timyates commented on starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java:
```diff
@@ -44,15 +53,35 @@ public String getDescription() {
+
+        if (generatorContext.isFeaturePresent(MicrometerFeature.class)) {
+            configuration.put("cassandra.default.advanced.metrics.factory.class", "MicrometerMetricsFactory");
+            configuration.addListItem("cassandra.default.advanced.metrics.session.enabled", "connected-nodes");
+            configuration.addListItem("cassandra.default.advanced.metrics.session.enabled", "cql-requests");
+            configuration.addListItem("cassandra.default.advanced.metrics.session.enabled", "bytes-sent");
+            configuration.addListItem("cassandra.default.advanced.metrics.session.enabled", "bytes-received");
+            configuration.addListItem("cassandra.default.advanced.metrics.node.enabled", "cql-requests");
+            configuration.addListItem("cassandra.default.advanced.metrics.node.enabled", "bytes-sent");
+            configuration.addListItem("cassandra.default.advanced.metrics.node.enabled", "bytes-received");
```
```suggestion
            configuration.put("cassandra.default.advanced.metrics.session.enabled", List.of("connected-nodes", "cql-requests", "bytes-sent", "bytes-received"));
            configuration.put("cassandra.default.advanced.metrics.node.enabled", List.of("cql-requests", "bytes-sent", "bytes-received"));
```

timyates commented on starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java:
```diff
@@ -44,15 +53,35 @@ public String getDescription() {
+            configuration.put("test-resources.containers.cassandra.exposed-ports[0].cassandra.port", 9042);
+            configuration.put("cassandra.default.basic.contact-points[0]", "localhost:${cassandra.port}");
+        } else {
+            configuration.put("cassandra.default.basic.contact-points[0]", "localhost:9042");
```
```suggestion
            configuration.put("cassandra.default.basic.contact-points", List.of("localhost:9042"));
```

timyates commented on starter-core/src/test/groovy/io/micronaut/starter/feature/database/CassandraSpec.groovy:
```diff
@@ -3,57 +3,83 @@ package io.micronaut.starter.feature.database
+        GeneratorContext commandContext = buildGeneratorContext([Cassandra.NAME])
+
+        then:
+        commandContext.configuration.get('cassandra.default.basic.contact-points[0]') == 'localhost:9042'
```
```suggestion
        commandContext.configuration.get('cassandra.default.basic.contact-points') == ['localhost:9042']
```

timyates commented on starter-core/src/test/groovy/io/micronaut/starter/feature/database/CassandraSpec.groovy:
```diff
@@ -3,57 +3,83 @@ package io.micronaut.starter.feature.database
+        GeneratorContext commandContext = buildGeneratorContext([Cassandra.NAME, TestResources.NAME])
+
+        then:
+        commandContext.configuration.get('cassandra.default.basic.contact-points[0]') == 'localhost:${cassandra.port}'
```
```suggestion
        commandContext.configuration.get('cassandra.default.basic.contact-points') == ['localhost:${cassandra.port}']
```

wetted committed:
Update starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java

Co-authored-by: Tim Yates <tim.yates@gmail.com>

wetted committed:
Update starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java

Co-authored-by: Tim Yates <tim.yates@gmail.com>

wetted committed:
Update starter-core/src/main/java/io/micronaut/starter/feature/database/Cassandra.java

Co-authored-by: Tim Yates <tim.yates@gmail.com>

wetted committed:
Update starter-core/src/test/groovy/io/micronaut/starter/feature/database/CassandraSpec.groovy

Co-authored-by: Tim Yates <tim.yates@gmail.com>

wetted committed:
Update starter-core/src/test/groovy/io/micronaut/starter/feature/database/CassandraSpec.groovy

Co-authored-by: Tim Yates <tim.yates@gmail.com>

wetted committed:
missing import

timyates reviewed (approved):
Generated a toml, properties and yaml app and they all worked ([when combined with the module pr](https://github.com/micronaut-projects/micronaut-cassandra/pull/377))

😎 👍

wetted commented:
reopened on #2126

timyates commented:
@wetted Don't forget to delete this [`cassandra+micrometer-updates`](https://github.com/micronaut-projects/micronaut-starter/tree/cassandra%2Bmicrometer-updates) branch once #2126 is merged 👍
